python/vast/utils/arrow.py

```diff
@@ -1,7 +1,125 @@
-import ipaddress
+import json
+import ipaddress as ip
+from typing import SupportsBytes
 
 import pyarrow as pa
 
+# Accepts a 128-bit buffer holding an IPv6 address and returns an IPv4 or IPv6
+# address.
+def unpack_ip(buffer: SupportsBytes) -> ip.IPv4Address | ip.IPv6Address:
+    num = int.from_bytes(buffer, byteorder="big")
+    # Convert IPv4 mapped addresses back to regular IPv4.
+    # https://tools.ietf.org/html/rfc4291#section-2.5.5.2
+    if (num >> 32) == 65535:
+        num = num - (65535 << 32)
+    return ip.ip_address(num)
+
+
+class AddressScalar(pa.ExtensionScalar):
+    def as_py(self) -> ip.IPv4Address | ip.IPv6Address:
+        return unpack_ip(self.value.as_py())
+
+
+class AddressType(pa.ExtensionType):
+    ext_name = "vast.address_type"
```

Comment:
```suggestion
    ext_name = "vast.address"
```

https://github.com/tenzir/vast/blob/1c3eec37e96db254ad73a405164718b4d435d779/libvast/include/vast/type.hpp#L756-L757